### PR TITLE
Support GEMM + Swiglu fused MLP

### PR DESCRIPTION
**Pull request by ksivaman**
# What does this PR do ?

This PR supports GEMM + Swiglu fused MLP via Transformer Engine sequential ops, including:

- TE `GroupedTensor` support in distributed optimizer.
- TE `GroupedTensor` checkpointing support for cases to load a multi-param (1 param per expert) checkpoint into a single param config.

**Note**: Support is already added in dev branch and this is cherry-picking #3890.

### Pre-checks

- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment the [@mcore-oncall](https://github.com/orgs/NVIDIA/teams/mcore-oncall) to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### CODE REVIEW

**Review by yaox12 — APPROVED** [review]

File: tests/unit_tests/models/test_mamba_moe_model.py [hunk 1]
```diff
@@ -287,6 +287,8 @@
     "moe_router_force_biased": None,
     "inference_grouped_gemm_backend": "auto",
     "inference_moe_disable_fused_quant_kernels": False,
+    "moe_mlp_glu_interleave_size": None,
```
Comment (yashaswikarnati) [inline tests/unit_tests/models/test_mamba_moe_model.py:290]: I wonder if we should add some test when  moe_mlp_glu_interleave_size is not None as this is something we are adding now ?

File: megatron/core/distributed/param_and_grad_buffer.py [hunk 2]
```diff
@@ -238,6 +238,9 @@ def __init__(
         # or bucket.grad_data.
         self.cached_param_buffer_shard_list = [None] * len(self.buckets)
         self.cached_grad_buffer_shard_list = [None] * len(self.buckets)
+        # Track grad mode used to create cached param views. Rebuild if mode changes to avoid
+        # mixing no_grad-created views with in-place updates in grad-enabled mode.
+        self._cached_param_buffer_shards_grad_enabled = None
```
Comment (deepakn94) [inline megatron/core/distributed/param_and_grad_buffer.py:248]: What is the purpose of this?

File: megatron/core/distributed/param_and_grad_buffer.py [hunk 3]
```diff
@@ -398,23 +401,28 @@ def start_param_sync(self, force_sync: bool = False):
...
-            ) as cm:
-                for idx, bucket in enumerate(self.buckets):
-                    if self.cached_param_buffer_shard_list[idx] is None:
-                        self.cached_param_buffer_shard_list[idx] = shard_buffer(
-                            bucket.param_data, self.intra_distributed_optimizer_instance_size
+            current_grad_enabled = torch.is_grad_enabled()
+            if self._cached_param_buffer_shards_grad_enabled != current_grad_enabled:
+                self.cached_param_buffer_shard_list = [None] * len(self.buckets)
+                self._cached_param_buffer_shards_grad_enabled = current_grad_enabled
+            with torch.no_grad():
```
Comment (deepakn94) [inline megatron/core/distributed/param_and_grad_buffer.py:413]: Comment on what this is trying to do?

File: megatron/core/optimizer/distrib_optimizer.py [hunk 4]
```diff
@@ -2459,8 +2526,29 @@ def _copy_main_params_to_model_params(self):
         if self.config.use_precision_aware_optimizer_no_fp8_or_ds_fp8:
             return
 
+        fp8_params, shard_fp32_from_fp8, shard_offsets_in_fp8 = (
```
Comment (deepakn94) [inline megatron/core/optimizer/distrib_optimizer.py:2529]: Comment on what this is trying to do?

File: megatron/core/transformer/moe/experts.py [hunk 5]
```diff
@@ -267,62 +290,191 @@ def _apply_bias(intermediate_parallel, bias_parallel, tokens_per_expert, permute
...
+            else self.linear_fc1.weight0.dtype
+        )
+        fc2_single_grouped_parameter = self.linear_fc2.single_grouped_parameter
+        fc2_weight_dtype = (
+            self.linear_fc2.weight.dtype
+            if fc2_single_grouped_parameter
+            else self.linear_fc2.weight0.dtype
+        )
+
+        # TODO:ksivamani: Why meta device?
```
Comment (timmoon10) [inline megatron/core/transformer/moe/experts.py:346]: @ksivaman If we didn't specify `device="meta"`, then we would allocate memory for the weights. However, we want to reuse the weights from the `GroupedLinear` module.

**Comment by zhongbozhu** [comment]
use_transformer_engine_op_fuser is not added to the arguments.py, is that intended?

**Comment by Connor-XY** [comment]
Close [this](https://github.com/NVIDIA/Megatron-LM/pull/3971) as we have https://github.com/NVIDIA/Megatron-LM/pull/4636
